Source/Layout/ASYogaLayoutSpec.mm

```diff
@@ -0,0 +1,156 @@
+  YGNODE_STYLE_SET_DIMENSION  (yogaNode, FlexBasis, style.flexBasis);
+
+  YGNodeStyleSetFlexDirection (yogaNode, yogaFlexDirection(style.flexDirection));
+  YGNodeStyleSetJustifyContent(yogaNode, yogaJustifyContent(style.justifyContent));
+  YGNodeStyleSetAlignSelf     (yogaNode, yogaAlignSelf(style.alignSelf));
+  ASStackLayoutAlignItems alignItems = style.alignItems;
+  if (alignItems != ASStackLayoutAlignItemsNotSet) {
+    YGNodeStyleSetAlignItems(yogaNode, yogaAlignItems(alignItems));
+  }
+
+  YGNodeStyleSetPositionType  (yogaNode, style.positionType);
+  ASEdgeInsets position = style.position;
+  ASEdgeInsets margin   = style.margin;
+  ASEdgeInsets padding  = style.padding;
+  ASEdgeInsets border   = style.border;
+
+  YGEdge edge = YGEdgeLeft;
+  for (int i = 0; i < YGEdgeAll + 1; ++i) {
+    YGNODE_STYLE_SET_DIMENSION_WITH_EDGE(yogaNode, Position, dimensionForEdgeWithEdgeInsets(edge, position), edge);
+    YGNODE_STYLE_SET_DIMENSION_WITH_EDGE(yogaNode, Margin, dimensionForEdgeWithEdgeInsets(edge, margin), edge);
+    YGNODE_STYLE_SET_DIMENSION_WITH_EDGE(yogaNode, Padding, dimensionForEdgeWithEdgeInsets(edge, padding), edge);
+    YGNODE_STYLE_SET_FLOAT_WITH_EDGE(yogaNode, Border, dimensionForEdgeWithEdgeInsets(edge, border), edge);
+    edge = (YGEdge)(edge + 1);
+  }
+
+  CGFloat aspectRatio = style.aspectRatio;
+  if (aspectRatio > FLT_EPSILON && aspectRatio < CGFLOAT_MAX / 2.0) {
+    YGNodeStyleSetAspectRatio(yogaNode, aspectRatio);
+  }
+
+  // For the root node, we use rootConstrainedSize above. For children, consult the style for their size.
+  if (parentYogaNode != NULL) {
+    YGNodeInsertChild(parentYogaNode, yogaNode, YGNodeGetChildCount(parentYogaNode));
+
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, Width, style.width);
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, Height, style.height);
+
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, MinWidth, style.minWidth);
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, MinHeight, style.minHeight);
+
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, MaxWidth, style.maxWidth);
+    YGNODE_STYLE_SET_DIMENSION(yogaNode, MaxHeight, style.maxHeight);
+
+    YGNodeSetMeasureFunc(yogaNode, &ASLayoutElementYogaMeasureFunc);
+  }
+
+  /* TODO(appleguy): STYLE SETTER METHODS LEFT TO IMPLEMENT
+   void YGNodeStyleSetOverflow(YGNodeRef node, YGOverflow overflow);
+   void YGNodeStyleSetFlex(YGNodeRef node, float flex);
+   */
+}
+
+- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)rootConstrainedSize
+{
+  YGNodeRef rootYogaNode = YGNodeNew();
+
+  // Apply the constrainedSize as a base, known frame of reference.
+  // If the root node also has style.*Size set, these will be overridden below.
+  // YGNodeCalculateLayout currently doesn't offer the ability to pass a minimum size (max is passed there).
+  YGNodeStyleSetMinWidth (rootYogaNode, yogaFloatForCGFloat(rootConstrainedSize.min.width));
+  YGNodeStyleSetMinHeight(rootYogaNode, yogaFloatForCGFloat(rootConstrainedSize.min.height));
```

Comment:
@nguyenhuy thanks for your comments here, it's very helpful to me! I just realized, what if this node has a min size already set on it - such as minWidth = 10% fraction? Do we have enough information here, maybe if we use the parentSize variant, to resolve that percentage and choose the more restrictive one?